### ISSUE TITLE
Repeated Struct typedef generates an error

### ISSUE THREAD

**Issue by sagamx** [eliben/pycparser#46]
This generates an error:

typedef const struct Ntf_FBlock_L_Type
{
    int FBlockIndex;
    int NumDev;
    int PtrPropTab;
} TNtfFBlockL, *pTNtfFBlockL;

typedef const struct Ntf_FBlock_L_Type
{
    int FBlockIndex;
    int NumDev;
    int PtrPropTab;
} TNtfFBlockL, *pTNtfFBlockL;

(it is repeated on purpose) However a repeated typedef like this:

typedef int INTEGER;
typedef int INTEGER;

does not generate any error nor this:

typedef const struct Ntf_FBlock_L_Type
{
    int FBlockIndex;
    int NumDev;
    int PtrPropTab;
} TNtfFBlockL;;

typedef const struct Ntf_FBlock_L_Type
{
    int FBlockIndex;
    int NumDev;
    int PtrPropTab;
} TNtfFBlockL;

So it just generates an error in the first case, it just throws a "ParserError before: ,"

What can I do to avoid this situation?

**Comment by eliben** [comment]
AFAICS, this generates an error in gcc as well:

error: redefinition of ‘struct Ntf_FBlock_L_Type’

**Comment by kirbyfan64** [comment]
@eliben This is valid in C11, though.

**Comment by natezb** [comment]
This issue is actually more general: any init-declarator-list that contains a name that's previously been typedef'd will fail to parse correctly, except if the name occurs as the first declarator, for which there are various productions that act as workarounds.

I've been thinking a bit about how to get the parsing right while using PLY's LALR(1) parser, and I believe I've finally figured it out.

The essential change is to allow a TYPEID to be part of a declarator. We introduce two categories of declarator: (1) an id-declarator (equivalent to what is currently just a 'declarator'), and (2) a typeid-declarator, which is identical but uses a TYPEID rather than an ID. In most cases, these two get recombined into a generic 'declarator', which can be either of the two. In the special case of a function definition, its declarator must be an id-declarator (this restriction is valid b/c a function name can't have been typedef'd already because such a typedef would have happened in the same outermost scope).

However, allowing TYPEIDs in declarators obviously will introduce some parsing conflicts. To fix these, the other main change is to make a restriction to declaration-specifiers (and specifier-qualifier-list) which reflects two things: (1) it must contain at least one type-specifier, and (2) if it contains a typedef-name, it may not contain any other type-specifiers. This is required so that the parser knows when to switch from parsing a specifiers list to parsing a declarator. It's also consistent with declaration-specifiers as described in section 6.7.2.2 of ISO/IEC 9899:TC3.

After making these changes,  you can then remove the existing productions that are workarounds to deal with incorrect parses.

In summary, the changes to the grammar include:
1. Allow TYPEIDs in declarators
2. Restrict declaration-specifiers and specifier-qualifier-list to contain at least one type-specifier, and _only_ one if it is a TYPEID
3. Remove now-unnecessary workaround productions

I have not yet implemented the actual code for each production yet—I wanted to check if these modifications to the grammar make sense and if you had any related suggestions. If everything seems OK, I'll move forward with putting together the code, running the tests, perhaps adding a few new tests, and issuing a pull request.

**Comment by natezb** [comment]
Upon further review it seems I accidentally deleted one of the parameter-declaration productions and restoring it reintroduced a few conflicts... but I still think it can probably be solved. I'll update when I have news.
